test/blackbox-tests/test-cases/pkg/convert-opam-commands.t

```diff
@@ -36,17 +36,34 @@ Generate a mock opam repository
   > install: [ make "install" ]
   > EOF
 
+  $ mkpkg variable-types <<EOF
+  > opam-version: "2.0"
+  > build: [
+  >   ["echo" local_var]
+  >   ["echo" _:explicit_local_var]
+  >   ["echo" foo:package_var]
+  >   ["echo" os-family]
+  > ]
+  > EOF
+
   $ solve_project <<EOF
   > (lang dune 3.8)
   > (package
   >  (name x)
-  >  (depends standard-dune))
+  >  (depends
+  >   standard-dune
+  >   variable-types))
   > EOF
   Solution for dune.lock:
   standard-dune.0.0.1
+  variable-types.0.0.1
   
 
   $ cat dune.lock/standard-dune.pkg
   (version 0.0.1)
   (install (run %{make} install))
-  (build (progn (run dune subst) (run dune build -p %{pkg:var:_:name} -j %{jobs} @install @runtest @doc)))
+  (build (progn (run dune subst) (run dune build -p %{pkg:name} -j %{jobs} @install @runtest @doc)))
+
+  $ cat dune.lock/variable-types.pkg
+  (version 0.0.1)
+  (build (progn (run echo %{pkg:local_var}) (run echo %{pkg:explicit_local_var}) (run echo %{pkg:foo:package_var}) (run echo %{os_family})))
```

Comment:
This is certainly an improvement as the `_` convention doesn't really fit with the rest of dune, but I think it's not a simple design. To see what I mean, try to expand these percent forms into the equivalent s-expressions. You would need something like overloading based on the number of arguments to express this and I don't think that's something we'd ever want to support.

To keep things simpler, I propose we do one of:

1. Pull package local variables into the current namespace as it was before. E.g. this would make it quite similar to our existing `%{deps}` or `%{targets}` vars.
2. Introduce a different form for "self" variables. Something like `%{pkg-self:..}`.

Another approach would be to just make `%{name}` first class, and add support for nesting e.g. `%{pkg:%{name}:...}`. I think this is "simple" (using your definition), but rather ugly.

Reply:
I'm in favour of option 2 as it means the `%{foo}` syntax can remain just for a fixed set of pform variables rather than arbitrarily-named package variables. So we'll have the `%{pkg:<package>:<variable>}` macro for variables from explicitly named packages and a new `%{pkg-self:<variable>}` macro for variables from the current package.